<launch>
    <!--指定USB摄像头设备编号-->
    <arg name="video_device" value="/dev/video0" />
    <!--画面宽度-->
    <arg name="image_width" value="640"/> 
    <!--画面高度-->
    <arg name="image_height" value="480" />
    <!--视频流编码-->
    <arg name="pixel_format" value="yuyv" />
    <!--设定摄像头的名字-->
    <arg name="camera_name" value="usb_cam"/> 
    <!--设定摄像头的Frame Id-->
    <arg name="camera_frame_id" value="usb_cam" />
    <!-- IO方法-->
    <arg name="io_method" value="mmap"/>
    <!--标定文件的路径-->
    <arg name="camera_info_url" value="package://aruco_mapping2/data/usb_cam.yaml"/>
    
    <!-- Aruco Marker的尺寸(宽高相同) 单位m-->
    <arg name="marker_size"      value="0.165"/>
    <!-- 基准码(世界坐标系)的ID-->
    <arg name="base_aruco_id" value="1"/>

   <node name="usb_cam" pkg="usb_cam" type="usb_cam_node" output="screen">
        <param name="video_device" value="$(arg video_device)" /> <!--指定USB摄像头设备编号-->
        <param name="image_width" value="$(arg image_width)" /> <!--画面宽度-->
        <param name="image_height" value="$(arg image_height)" /> <!--画面高度-->
        <param name="pixel_format" value="$(arg pixel_format)" /> <!--视频流编码-->
        <param name="camera_name" value="$(arg camera_name)"/> <!--设定摄像头的名字-->
        <param name="camera_frame_id" value="$(arg camera_frame_id)" /> <!--设定摄像头的Frame Id-->
        <param name="io_method" value="$(arg io_method)"/>
        <param name="camera_info_url" value="$(arg camera_info_url)"/>
    </node>
    <node ns="usb_cam" name="usb_cam_image_proc" pkg="image_proc" type="image_proc"/>

    <!--Marker 参考Frame-->
    <!--<arg name="ref_frame"       default="base"/>--> 
    <!-- leave empty and the pose will be published wrt param parent_name -->
    <arg name="ref_frame"       default=""/>

    <node pkg="aruco_ros" type="marker_publisher" name="aruco_marker_publisher">
        <remap from="/camera_info" to="/usb_cam/camera_info" />
        <remap from="/image" to="/usb_cam/image_rect_color" />
        <param name="image_is_rectified" value="True"/>
        <param name="marker_size"        value="$(arg marker_size)"/>
        <param name="reference_frame"    value="$(arg ref_frame)"/>   <!-- frame in which the marker pose will be refered -->
        <param name="camera_frame"       value="usb_cam"/>
    </node>
    <!-- 视频预览框 -->
    <node name="image_view" pkg="image_view" type="image_view" respawn="false" output="screen">
        <remap from="image" to="/aruco_marker_publisher/result"/>
        <!--Image必须要被Remap 暂时不清楚其机制-->
        <!--修改为 /camera/image_raw 之后不能正确运行-->
        <!--<param name="image" value="/camera/image_raw" />-->
        <param name="autosize" value="true" />
    </node>
    <!--可视化-->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find aruco_mapping2)/rviz/aruco_local.rviz" />
    <!-- Py 文件 type要添加.py文件后缀 -->
    <!-- 绘制Aruco地图 -->
    <node name="draw_aruco_map" pkg="aruco_mapping2" type="draw_aruco_map.py" />
    <!-- 相机位姿优化 -->
    <node name="camera_pose_optimization" pkg="aruco_mapping2" type="camera_pose_optimization.py" />
</launch>

